spark2/src/test/java/org/apache/iceberg/spark/source/TestSparkTableUtil.java

```diff
@@ -298,4 +309,69 @@ public void testImportWithNameMappingForVectorizedParquetReader() throws Excepti
 
     Assert.assertEquals(expected.stream().map(SimpleRecord::getData).collect(Collectors.toList()), actual);
   }
+
+  @Test
+  public void testAvroReaderWithNameMapping() throws IOException {
+    File avroFile = temp.newFile();
+    org.apache.avro.Schema avroSchema = SchemaBuilder.record("TestRecord")
+        .namespace("org.apache.iceberg.spark.data")
+        .fields()
+        .requiredInt("id")
+        .requiredString("name")
+        .endRecord();
+
+    GenericRecord record1 = new GenericData.Record(avroSchema);
+    record1.put("id", 1);
+    record1.put("name", "Bob");
+
+    GenericRecord record2 = new GenericData.Record(avroSchema);
+    record2.put("id", 2);
+    record2.put("name", "Alice");
+
+    DatumWriter<GenericRecord> datumWriter = new GenericDatumWriter<>(avroSchema);
+    DataFileWriter<GenericRecord> dataFileWriter = new DataFileWriter<>(datumWriter);
+
+    dataFileWriter.create(avroSchema, avroFile);
+    dataFileWriter.append(record1);
+    dataFileWriter.append(record2);
+    dataFileWriter.close();
+
+    DataFile avroDataFile = DataFiles.builder(PartitionSpec.unpartitioned())
+        .withFormat("avro")
+        .withFileSizeInBytes(avroFile.length())
+        .withPath(avroFile.getAbsolutePath())
+        .withRecordCount(2)
+        .build();
+
+    Schema filteredSchema = new Schema(
+        required(1, "name", Types.StringType.get())
+    );
+
+    NameMapping nameMapping = MappingUtil.create(filteredSchema);
```

Comment:
I like that `filteredSchema` is used to create the mapping. Could you use a full schema for the table (1: name, 2: id) so that we can verify that the ID is projected as null?

Reply:
We can not use a full schema for the table because it cannot build Avro read schema (with `name` and `id` column) by using pruned schema (with `name` column) in `ProjectionDatumReader#setSchema`.

Also, `setSchema` will ignore the name mapping if the Avro file schema has Ids. So we might need to utilize the `RemoveIds` from the Avro test via moving it to source dir and changing it to the public class. What do you think about this?

Reply:
Can you clarify what you mean by "it cannot build Avro read schema by using pruned schema"? We should be able to use any table schema here, and without a mapping for id, it should be ignored.

We added special handling for this case where the table schema and the file schema use the same name, but the column should not be projected because this ID is missing.

> setSchema will ignore the name mapping if the Avro file schema has Ids

Isn't this why the Avro file schema is created without IDs and no Iceberg classes are used for the write? I don't see why we would have IDs in the Avro file.

Reply:
This is the stack trace that when using a full schema for the table:

>
org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 0.0 failed 1 times, most recent failure: Lost task 0.0 in stage 0.0 (TID 0, localhost, executor driver): java.lang.IllegalArgumentException: Missing required field: id 
at org.apache.iceberg.relocated.com.google.common.base.Preconditions.checkArgument(Preconditions.java:217)
at org.apache.iceberg.avro.BuildAvroProjection.record(BuildAvroProjection.java:96)
at org.apache.iceberg.avro.BuildAvroProjection.record(BuildAvroProjection.java:41)
at org.apache.iceberg.avro.AvroCustomOrderSchemaVisitor.visit(AvroCustomOrderSchemaVisitor.java:51) 
at org.apache.iceberg.avro.AvroSchemaUtil.buildAvroProjection(AvroSchemaUtil.java:104)
at org.apache.iceberg.avro.ProjectionDatumReader.setSchema(ProjectionDatumReader.java:60)
at org.apache.avro.file.DataFileStream.initialize(DataFileStream.java:132)
at org.apache.avro.file.DataFileReader.<init>(DataFileReader.java:106)
at org.apache.avro.file.DataFileReader.<init>(DataFileReader.java:98)
at org.apache.avro.file.DataFileReader.openReader(DataFileReader.java:66)                                                                                                                                          at org.apache.iceberg.avro.AvroIterable.newFileReader(AvroIterable.java:94)
at org.apache.iceberg.avro.AvroIterable.iterator(AvroIterable.java:77)
at org.apache.iceberg.spark.source.RowDataReader.open(RowDataReader.java:106)                                                                      at org.apache.iceberg.spark.source.BaseDataReader.next(BaseDataReader.java:73)
at org.apache.spark.sql.execution.datasources.v2.DataSourceRDD$$anon$1.hasNext(DataSourceRDD.scala:49)
at org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37) 
at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
...

So that I didn't use the full schema and assert the id column.

>Isn't this why the Avro file schema is created without IDs and no Iceberg classes are used for the write? I don't see why we would have IDs in the Avro file.

Yes, agree to create Avro file without IDs. Here I just want to use `RemoveIds` class from Iceberg.